### PR TITLE
➕(core) replace json schema validation with pydantic model

### PR DESCRIPTION
**Pull request by quitterie-lcs**
# Purpose

[pydantic](https://pydantic-docs.helpmanual.io/) seems to be a better solution for `marion`'s schema validation.
It is a more pythonic architecture to manipulate context data. For example, date can be manipulated as `datetime` python object and not as string anymore.

# Proposal 
All the JSON schema definitions have been replaced by pydantic models in issuers.

### CODE REVIEW

File: src/howard/howard/issuers/realisation.py [hunk 1]
```diff
@@ -18,111 +28,121 @@ class CertificateScope(Enum):
...
+class Date(BaseModel):
+    """Timeframe model definition"""
+
+    start: datetime = ...
+    end: datetime = ...
+
+    class Config:
+        """Timeframe model configuration"""
+
+        title = "date"
```
Comment (jmaupetit) [inline src/howard/howard/issuers/realisation.py:40]: If we only set the model's `title` in its configuration, I think we can remove the whole configuration as it's only used for JSON schema generation.

File: src/howard/howard/templates/marion/certificates/realisation.html [hunk 2]
```diff
@@ -63,8 +63,8 @@ <h1>CERTIFICAT DE REALISATION</h1>
           </div>
           <div class="dates">
             qui s'est déroulée du
-            <span>{{ course.session.date.from }}</span> au
-            <span>{{ course.session.date.to }} </span>
+            <span>{{ course.session.date.start }}</span> au
+            <span>{{ course.session.date.end }} </span>
```
Comment (jmaupetit) [inline src/howard/howard/templates/marion/certificates/realisation.html:67]: Once deserialized (thanks to pydantic) and typed, one should use Django's `date` filter :tada: 

```suggestion
            <span>{{ course.session.date.end | date }} </span>
```

File: src/howard/howard/issuers/realisation.py (hunk 1 above)
Comment (jmaupetit) [reply src/howard/howard/issuers/realisation.py:40]: Should we also set `extra` instead of removing the configuration?

Comment (quitterie-lcs) [reply src/howard/howard/issuers/realisation.py:76]: I agree with your first remark. As I began to implement still using JSON schema from pydantic, I did not noticed that it was useless now that the implementation is free of JSON.

Comment (quitterie-lcs) [reply src/howard/howard/issuers/realisation.py:76]: You mean you want that no extra fields are allowed?

Comment (jmaupetit) [reply src/howard/howard/issuers/realisation.py:76]: Not sure if we have to be pedantic or not :wink:

File: CHANGELOG.md [hunk 3]
```diff
@@ -10,9 +10,14 @@ Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to
...
 
+- Add `pydantic` package for API/schema validation
 - Add `howard` package that bundles fun-specific certificates
 - Add `CertificateRequest` model, serializer and API viewset
 - Add `DummyCertificate` example certificate issuer
 - Implement certificate issuer pattern
 
+### Removed
+
+- Remove `JSON` schema validation
```
Comment (jmaupetit) [inline CHANGELOG.md:21]:
```suggestion
- Pure `JSON` schema validation has been deprecated
```

File: src/howard/howard/issuers/realisation.py [hunk 4]
```diff
@@ -18,126 +28,168 @@ class CertificateScope(Enum):
...
-                    "additionalProperties": False,
-                },
-                "organization": {"$ref": "#/definitions/organization"},
-            },
-            "required": ["name", "session", "organization"],
-        },
-    }
-}
+class Date(BaseModel):
+    """Timeframe model definition"""
```
Comment (jmaupetit) [inline src/howard/howard/issuers/realisation.py:32]:
```suggestion
class DateSpan(BaseModel):
    """DateSpan model definition"""
```

File: src/howard/howard/issuers/realisation.py [hunk 5]
```diff
@@ -18,126 +28,168 @@ class CertificateScope(Enum):
...
+            try:
+                return datetime.strptime(value, "%d/%m/%Y")
+            except ValueError as err:
+                logging.error(
+                    "Input end date '%s' does not match '%s' date format.",
+                    value,
+                    short_date_format,
+                )
+                logging.debug("Raised error was: %s", err)
+        return value
```
Comment (jmaupetit) [inline src/howard/howard/issuers/realisation.py:71]: I would factorize this in a single private class method to avoid duplication.

File: src/howard/howard/issuers/realisation.py [hunk 6]
```diff
@@ -18,126 +28,168 @@ class CertificateScope(Enum):
...
+
+    start: datetime = ...
+    end: datetime = ...
+
+    # pylint: disable=no-self-use
+    # pylint: disable=no-self-argument
+    @validator("start", pre=True)
+    def parse_start(cls, value):
+        """Validates start date format when the value is given as a string."""
+        short_date_format = "%d/%m/%Y"
```
Comment (jmaupetit) [inline src/howard/howard/issuers/realisation.py:42]: I thing you could you the corresponding Django setting instead.

File: src/howard/howard/issuers/realisation.py [hunk 7]
```diff
@@ -18,126 +28,168 @@ class CertificateScope(Enum):
...
+        if isinstance(value, str):
+            try:
+                return datetime.strptime(value, "%d/%m/%Y")
+            except ValueError as err:
+                logging.error(
```
Comment (jmaupetit) [inline src/howard/howard/issuers/realisation.py:47]: You should configure a logger first in the header and use this logger. See https://github.com/openfun/ralph/blob/master/src/ralph/parsers.py#L11 for example.

File: src/howard/howard/issuers/realisation.py [hunk 8]
```diff
@@ -18,126 +28,168 @@ class CertificateScope(Enum):
...
+    context_model = ContextModel
+    context_query_model = ContextQueryModel
+
     def fetch_context(self, **context_query):
         """Fetch the context that will be used to compile the certificate template."""
 
-        short_date_format = "%d/%m/%Y"
-
-        self.validate_context_query(context_query)
+        context_query_instance = self.validate_context_query(context_query)
```
Comment (jmaupetit) [inline src/howard/howard/issuers/realisation.py:185]:
```suggestion
        validated = self.validate_context_query(context_query)
```

File: src/marion/marion/certificates/issuers/base.py [hunk 9]
```diff
@@ -134,29 +134,31 @@ def __init__(self, identifier=None):
...
 
-        if cls.context_query_schema is None:
+        if cls.context_query_model is None:
             raise CertificateIssuerContextQueryValidationError(
-                str(_("Context query schema is missing"))
+                str(_("Context query model is missing"))
             )
         try:
-            validate(context_query, cls.context_query_schema)
+            # pylint: disable=not-callable
```
Comment (jmaupetit) [inline src/marion/marion/certificates/issuers/base.py:160]: Maybe add a type hint to the `context_model` and `context_query_model` class attributes to avoid having to disable this pylint error.

File: src/marion/marion/certificates/issuers/dummy.py [hunk 10]
```diff
@@ -37,9 +51,10 @@ def fetch_context(self, **context_query):
 
         """
 
-        self.validate_context_query(context_query)
-        fullname = context_query.get("fullname", None)
-        return Context({"fullname": fullname, "identifier": self.identifier})
+        context_query_instance = self.validate_context_query(context_query)
```
Comment (jmaupetit) [inline src/marion/marion/certificates/issuers/dummy.py:54]:
```suggestion
        validated = self.validate_context_query(context_query)
```

File: src/marion/marion/certificates/tests/test_models.py [hunk 11]
```diff
@@ -6,73 +6,80 @@
...
+        DjangoFieldError, match="A pydantic model is missing for the 'data' field."
     ):
         instance.full_clean()
 
     # With a schema attribute
+    class PydanticModelA(BaseModel):
+        fullname: str = ...
+
+        class Config:
+            title = "pydantic_model_a"
```
Comment (jmaupetit) [inline src/marion/marion/certificates/tests/test_models.py:34]: required?

File: src/marion/marion/certificates/tests/test_serializers.py [hunk 12]
```diff
@@ -3,90 +3,96 @@
...
+        "A pydantic model is missing for the 'misc' field."
+        in instance.errors.get("misc")[0]
+    )
+
+    # With a pydantic model attribute
+    class PydanticModelA(BaseModel):
+        fullname: str = ...
+
+        class Config:
+            title = "pydantic_model_a"
```
Comment (jmaupetit) [inline src/marion/marion/certificates/tests/test_serializers.py:38]: required?

File: src/howard/howard/issuers/realisation.py (hunk 1 above)
Comment (quitterie-lcs) [reply src/howard/howard/issuers/realisation.py:76]: With the `JSON` schema implementation, we did not specify to forbid additional properties, so we can keep the same position also with `pydantic`.

File: src/marion/marion/certificates/tests/test_models.py (hunk 11 above)
Comment (quitterie-lcs) [reply src/marion/marion/certificates/tests/test_models.py]: Nope

File: src/marion/marion/certificates/tests/test_serializers.py (hunk 12 above)
Comment (quitterie-lcs) [reply src/marion/marion/certificates/tests/test_serializers.py]: Nope

File: src/howard/howard/issuers/realisation.py (hunk 1 above)
Comment (jmaupetit) [reply src/howard/howard/issuers/realisation.py]: Agreed.

File: src/marion/marion/certificates/issuers/base.py (hunk 9 above)
Comment (quitterie-lcs) [reply src/marion/marion/certificates/issuers/base.py:160]: I removed it for now but the point is that, I have to define it as `None` at the beginning, so it will be a `NoneType` in any case and consequently this `lint` error is raised. WDYT?

File: src/howard/howard/templates/marion/certificates/realisation.html [hunk 13]
```diff
@@ -87,7 +87,7 @@ <h1>CERTIFICAT DE REALISATION</h1>
               Fait à : <span>{{ course.organization.location }}</span>
             </div>
             <div class="date">
-              Le : <span>{{ creation_date }}</span>
+              Le : <span>{{ creation_date | date:"d/m/Y" }}</span>
```
Comment (quitterie-lcs) [inline src/howard/howard/templates/marion/certificates/realisation.html:90]: @jmaupetit `creation_date` and `delivery_stamp` are both the same actually. They come form `self.created`. However, this variable is instanciated as an isoformat. Is it really necessary? In the template, we need to reformat the creation_date to get a short format, and for it , we need to go back to a datetime from an isoformat. Wouldn't it be better to remove the `isoformat()` instanciation in the base class? What do you think?

File: src/howard/howard/issuers/realisation.py [hunk 14]
```diff
@@ -1,143 +1,159 @@
...
+    FEMALE = "Mme"
+
+
+class CertificateScope(str, Enum):
+    """Allowed scopes for the RealisationCertificate."""
 
     FORMATION = "action de formation"
     BILAN = "bilan de compétences"
     VAE = "action de VAE"
     APPRENTISSAGE = "action de formation par apprentissage"
```
Comment (jmaupetit) [inline src/howard/howard/issuers/realisation.py:32]: Why not adding this to enums (also for gender), so that we are not forced to use `scope.value` but `scope` in the templates?
```suggestion
    def __str__(self):
        return f"{self.value}"
```

File: src/howard/howard/templates/marion/certificates/realisation.html [hunk 15]
```diff
@@ -43,7 +43,7 @@ <h1>CERTIFICAT DE REALISATION</h1>
         </section>
         <section id="student">
           <div class="name">
-            <span class="gender">{{ student.gender }}</span>
+            <span class="gender">{{ student.gender.value }}</span>
```
Comment (jmaupetit) [inline src/howard/howard/templates/marion/certificates/realisation.html:46]: See Enum objects suggestion:

```suggestion
            <span class="gender">{{ student.gender }}</span>
```

File: src/howard/howard/templates/marion/certificates/realisation.html [hunk 16]
```diff
@@ -58,13 +58,13 @@ <h1>CERTIFICAT DE REALISATION</h1>
           <div class="scope">
             <p>
               Nature de l'action de formation :
-              <span>{{ course.session.scope }}</span>
+              <span>{{ course.session.scope.value }}</span>
```
Comment (jmaupetit) [inline src/howard/howard/templates/marion/certificates/realisation.html:61]: Ditto

```suggestion
              <span>{{ course.session.scope }}</span>
```

File: src/howard/howard/templates/marion/certificates/realisation.html (hunk 13 above)
Comment (jmaupetit) [reply src/howard/howard/templates/marion/certificates/realisation.html:90]: Yes, you are right: it should be a `datetime` instance instead.

File: src/marion/marion/certificates/issuers/base.py [hunk 17]
```diff
@@ -134,29 +134,31 @@ def __init__(self, identifier=None):
...
-        if cls.context_schema is None:
+        if cls.context_model is None:
             raise CertificateIssuerContextValidationError(
-                str(_("Context schema is missing"))
+                str(_("Context model is missing"))
             )
         try:
-            validate(context, cls.context_schema)
+            # pylint: disable=not-callable
+            return cls.context_model(**context)
```
Comment (jmaupetit) [inline src/marion/marion/certificates/issuers/base.py:145]: Maybe adding type hints in l.119-120 would help pylint here?

File: src/marion/marion/certificates/issuers/base.py (hunk 9 above)
Comment (jmaupetit) [reply src/marion/marion/certificates/issuers/base.py:160]: I'm not sure to understand: defining types as suggested (or NoneType) does or does not raise this linting error?

File: src/marion/marion/certificates/issuers/base.py (hunk 17 above)
Comment (quitterie-lcs) [reply src/marion/marion/certificates/issuers/base.py:145]: The point is that I can add the type hint `ContextModel` but I have to assign the variable to `None` at first... The pylint error is still raised

File: src/marion/marion/certificates/issuers/base.py (hunk 9 above)
Comment (quitterie-lcs) [reply src/marion/marion/certificates/issuers/base.py:160]: With this definition, it still raises linting error

File: src/howard/howard/issuers/realisation.py [hunk 18]
```diff
@@ -1,143 +1,163 @@
...
-    """Allowed scopes for the RealisationCertificate"""
+
+class Gender(str, Enum):
+    """Gender definitions."""
+
+    MALE = "Mr"
+    FEMALE = "Mme"
+
+    def __str__(self):
+        return f"{self.value}"
```
Comment (jmaupetit) [inline src/howard/howard/issuers/realisation.py:26]: What about factoring a bit?

```suggestion
class StrEnum(str, Enum):
    """String Enum."""

    def __str__(self):
        return f"{self.value}"


class Gender(StrEnum):
    """Gender definitions."""

    MALE = "Mr"
    FEMALE = "Mme"
```

File: src/howard/howard/issuers/realisation.py [hunk 19]
```diff
@@ -1,143 +1,163 @@
...
+            return value.date
+
+        raise ValidationError("Date value must be of type int, str or datetime.date.")
+
+
+class DateSpan(BaseModel):
+    """DateSpan model definition."""
+
+    start: ArrowDate = ...
+    end: ArrowDate = ...
```
Comment (jmaupetit) [inline src/howard/howard/issuers/realisation.py]: As discussed IRL, we decided to remove the ellipsis to maintain the mypy compatibility.

File: src/howard/howard/tests/issuers/test_realisation.py [hunk 20]
```diff
@@ -1,41 +1,480 @@
...
+            },
+            "none is not an allowed value",
+        ),
+        (
+            {
+                "name": "edx",
+                "manager": "CEO, Agarwal Anant",
+                "location": "Boston, MA",
+            },
+            "value is not a valid dict",
```
Comment (jmaupetit) [inline src/howard/howard/tests/issuers/test_realisation.py:188]: I would also test an empty dict or dicts with missing required keys (one key at a time), _e.g._ `name`.

File: src/howard/howard/tests/issuers/test_realisation.py [hunk 21]
```diff
@@ -1,41 +1,480 @@
...
+            {
+                "datespan": {
+                    "start": ArrowDate("2021/01/01"),
+                    "end": ArrowDate("2021/12/31"),
+                },
+                "duration": 23,
+                "scope": "formation",
+                "manager": "CEO Agarwal Anant",
+            },
+            "value is not a valid dict",
```
Comment (jmaupetit) [inline src/howard/howard/tests/issuers/test_realisation.py:334]: The invalid field name is not mentioned in pydantic's errors?

File: src/howard/howard/issuers/realisation.py (hunk 18 above)
Comment (quitterie-lcs) [reply src/howard/howard/issuers/realisation.py:26]: Smart ;)
